Chapter 8: Resource Bindings (#5)
amolenk/docs: docs/architecture/dapr-for-net-developers/bindings-buildingblock.md

![Output binding](media/bindings-output.png)

**Figure 8-3**. Dapr output binding.

1. The Dapr sidecar reads the binding configuration file with the information on how to connect to the external resource. In the example, the external resource is a Twilio SMS account.
2. Your application invokes the `/v1.0/bindings/sms` endpoint on the Dapr sidecar. In this case, it uses an HTTP POST to invoke the API. It's also possible to use gRPC.
3. The binding component running in the Dapr sidecar calls the external messaging system to send the message. The message will contain the payload passed in the POST request.

As an example, you can invoke an output binding by invoking the Dapr API using curl:

```bash
curl -X POST -H "Content-Type: application/json" -d payload.json http://localhost:3500/v1.0/bindings/sms
```

> This is the content of the `payload.json` file:

Reviewer: I think it reads better to combine the command and payload in a single example:
```
curl -X POST http://localhost:3500/v1.0/state/<store_name> \
  -H "Content-Type: application/json" \
  -d '[
        {
          "key": "Hello",
          "value": "World",
          "options": {
            "consistency": "strong"
          }
        }
      ]' 
```
The example above is from the state management chapter where I've got multiple curl snippets to explain how it works. Unfortunately, there isn't a cross platform way to run multi-line curl commands. I'm using the '\' syntax which works in bash (and thus Git bash on Windows), because the Dapr docs also use that syntax.

Consider, for example, a Twitter account that triggers an event whenever a user tweets a keyword. Your service exposes an event handler that receives and processes the tweet. Once complete, your service triggers an event that invokes an external Twilio service. Twilio sends an SMS message that includes the tweet. Figure 8-1 show the conceptual architecture of this operation.

![Input binding](media/resource-binding-conceptual-architecture.png)

**Figure 8-1**. Conceptual architecture of a Dapr resource binding.

At first glance, resource binding behavior may appear similar to the [Publish/Subscribe pattern](publish-subscribe-buildingblock.md) described earlier in this book. While they share similarities, there are differences. Publish/subscribe focuses on asynchronous communication between Dapr services. Resource binding has a much wider scope. It focuses on system interoperability across software platforms. Exchanging information between disparate applications, datastores, and services outside your microservice application.

> Lastly, Dapr resource bindings can be easily swapped at runtime without code changes.

Reviewer: I'd move this info to the end of the first paragraph, where we already talk about components.

Reviewer 2: That's really good suggestion. Makes more sense after the componenets discussion. Done!

## How it works

> Dapr resource binding starts with a yaml-based file called a *binding configuration*. This file describes the resource to which you'll communicate along with its binding details. Once configured, your service can receive events from the resource or trigger events on it. (Binding configurations are presented in detail later in the *Components* section.)

Reviewer: ```suggestion
Dapr resource binding starts with a component configuration file. This yaml file describes the type of resource to which you'll bind along with its configuration settings. Once configured, your service can receive events from the resource or trigger events on it.

> [!NOTE]
> Binding configurations are presented in detail later in the *Components* section.
```

### Input bindings

Input bindings trigger your code with incoming events from external resources. To receive events and data, you register a public endpoint from your service that becomes the *event handler*. Figure 8-2 shows the architecture:

![Input binding](media/bindings-input.png)

**Figure 8-2**. Dapr input binding.

Figure 8.2 describes the steps for receiving events from an external Twitter account:

> 1. The Dapr sidecar reads the binding configuration file and subscribes to the event specified for external resource. In the example, the event source is a Twitter account.

Reviewer: ```suggestion
1. The Dapr sidecar reads the binding configuration file and subscribes to the event specified for the external resource. In the example, the event source is a Twitter account.
```

> 2. When a matching Tweet is published on Twitter, the Dapr runtime triggers an event.

Reviewer: Steps 2/3 read like there's an additional Dapr runtime component that isn't displayed in the diagram. Think we should change step 2 to something like:
"When a (...) Twitter, the binding component running in the Dapr sidecar triggers an event."

3. The Dapr sidecar responds by invoking the endpoint (that is, event handler) configured for the binding. In the example, the service listens for an HTTP POST on the `/tweet` endpoint on port 6000. Because it's an HTTP POST operation, the JSON payload for the event is passed in the request body.
4. After handling the event, the service returns an HTTP status code `200 OK`.

The following ASP.NET Core controller provides an example of handling an event triggered by the Twitter binding:

```csharp
[ApiController]
public class SomeController : ControllerBase
{
    [HttpPost("/tweet")]
    public ActionResult Post(JsonElement data)
    {
        // handle tweet

> 

Reviewer: ```suggestion
```

        //...

        // acknowledge message
        return Ok();
    }
  
    //...  
}
```

> If the operation should error, you would return the appropriate 400 or 500 level HTTP status code. For bindings that feature *at-least-once-delivery* guarantees, the Dapr sidecar will retry the trigger. Check out [the documentation of the different bindings](https://github.com/dapr/docs/tree/master/concepts/bindings) to see whether they offer *at-least-once* or *exactly-once* delivery guarantees.

Reviewer: ```suggestion
If the operation should error, you would return the appropriate 400 or 500 level HTTP status code. For bindings that feature *at-least-once-delivery* guarantees, the Dapr sidecar will retry the trigger. Check out [the documentation of the different bindings](https://docs.dapr.io/operations/components/setup-bindings/supported-bindings/) to see whether they offer *at-least-once* or *exactly-once* delivery guarantees.
```

> If the operation should error, you would return the appropriate 400 or 500 level HTTP status code. For bindings that feature *at-least-once-delivery* guarantees, the Dapr sidecar will retry the trigger. Check out [the documentation of the different bindings](https://github.com/dapr/docs/tree/master/concepts/bindings) to see whether they offer *at-least-once* or *exactly-once* delivery guarantees.

Reviewer: Changed the URL to the new docs.dapr.io. Cannot find the info on at-least-once vs exactly-once though. (Is it really exactly-once or at-most-once? What if the sidecar crashes after receiving an external trigger, but before delivering it to the app?)

### Output bindings

Dapr also includes *output binding* capabilities. They enable your service to trigger an event that invokes an external resource. Again, you start by configuring a binding configuration yaml file that describes the output binding. Once in place, you trigger an event that invokes the bindings API on the Dapr sidecar of your application. Figure 8-3 shows the architecture of an output binding:

![Output binding](media/bindings-output.png)

**Figure 8-3**. Dapr output binding.

1. Your application invokes the `/v1.0/bindings/sms` endpoint on the Dapr sidecar. In this case, it uses an HTTP POST to invoke the API. It's also possible to use gRPC.
2. The Dapr sidecar calls the external messaging system to send the message. The message will contain the payload passed in the POST request.

> As an example, the following method implements an output binding in a ASP.NET Core application:

Reviewer: ```suggestion
As an example, the following method implements an output binding in an ASP.NET Core application:
```

```c#

> private async Task SendSMSAsync(IHttpClientFactory clientFactory)

Reviewer: Maybe change this example to a curl call to keep all .NET samples use the .NET SDK?

{
    var payload = new
    {
        data = "Welcome to this awesome service",
        metadata = {

>           toNumber = "+31612345678"

Reviewer: ```suggestion
          toNumber = "555-3277"
```

Reviewer: Can you spot the easter egg? ;-)

        },
        operation = "create"
    };

    // send message using output-binding
    var content = new StringContent(
        JsonSerializer.Serialize(payload),
        Encoding.UTF8, "application/json");

    var httpClient = clientFactory.CreateClient();

    await httpClient.PostAsync("http://localhost:3500/v1.0/bindings/sms", content);
}
```

> To start, note how the previous figure uses a standard `HttpClient` object to execute an HTTP POST operation. There are no references to a specific Dapr library or SDK. Note also that the port is the same as used by the Dapr sidecar (in this case, the default HTTP port `3500`).

Reviewer: ```suggestion
To start, note how the previous figure uses a standard `HttpClient` object to execute an HTTP POST operation. There are no references to a specific Dapr library or SDK. Note also that the port is the same as used by the Dapr sidecar (in this case, the default Dapr HTTP port `3500`).
```

The structure of the payload (that is, message sent) will vary per binding. In the example above, the payload contains a `data` element  with a message. Bindings to other types of external resources can be different, especially for the metadata that is sent. Each payload must also contain an `operation` field, that defines the operation the binding will execute. The above example specifies a  `create` operation that creates the SMS message. Common operations include:

- create
- get
- delete
- list

The documentation for each binding describes the available operations and how to invoke them.

## Using the .NET SDK

> The Dapr .NET SDK provides language specific support for .NET Core developers. In the following example, the call to the `HttpClient.PostAsync()` is replaced with the `DaprClient.InvokeBindingAsync()` method. This specialized method simplifies invoking a configured output resource:

Reviewer: ```suggestion
The Dapr .NET SDK provides language specific support for .NET Core developers. In the following example, the call to the `HttpClient.PostAsync()` is replaced with the `DaprClient.InvokeBindingAsync()` method. This specialized method simplifies invoking a configured output binding:
```

```csharp
private async Task SendSMSAsync([FromServices] DaprClient daprClient)
{
    var string message = "Welcome to this awesome service";

>     var metadata = new Dictionary<string,string> { { "toNumber", "+31612345678"} };

Reviewer: ```suggestion
    var metadata = new Dictionary<string,string> { { "toNumber", "555-3277"} };
```

    await DaprClient.InvokeBindingAsync("sms", "create", message, metadata);
}
```

The method expects the `metadata` and `message` values.

When used to invoke a binding, the `DaprClient` uses gRPC to call the Dapr API on the Dapr sidecar.

### Binding Components

Under the hood, resource bindings are implemented with Dapr binding components. They're contributed by the community and written in Go. If you need to integrate with an external resource for which no Dapr binding exists yet, you can create it yourself. Check out the [Dapr components-contrib repo](https://github.com/dapr/components-contrib) to see how you can contribute a binding.

> > [!NOTE]

Reviewer: Do we want to repeat this note in each chapter or move it to one of the intro chapters?

> Dapr and all of its components are written in the [Golang](https://golang.org/) (Go) language. Go is considered a modern, cloud-native programming platform.

You configure bindings using a yaml configuration file. Here's an example configuration for the Twitter binding:

```yaml
apiVersion: dapr.io/v1alpha1
kind: Component
metadata:
  name: twitter-mention
  namespace: default
spec:
  type: bindings.twitter
  metadata:
  - name: consumerKey
    value: "****" # twitter api consumer key, required
  - name: consumerSecret
    value: "****" # twitter api consumer secret, required
  - name: accessToken
    value: "****" # twitter api access token, required
  - name: accessSecret
    value: "****" # twitter api access secret, required
  - name: query
    value: "dapr" # your search query, required
```

Each binding configuration contains a general `metadata` element with a `name` and `namespace` field. Dapr will determine the endpoint to invoke your service based upon the configured `name` field. In the above example, Dapr will invoke the method annotated with `/twitter-mention` in your service when an event occurs.

> In the `spec` element, you specify the `type` of the binding along with binding specific `metadata`. The example specifies credentials for accessing a Twitter account using its API. The metadata can differ between input and output bindings. For example, to use the Twitter as an input binding, you need to specify the text to search for in tweets using the `query` field. Every time a matching tweet is sent, the Dapr sidecar will invoke the `/twitter-mention` endpoint on the service. It will also deliver the contents of the Tweet

Reviewer: ```suggestion
In the `spec` element, you specify the `type` of the binding along with binding specific `metadata`. The example specifies credentials for accessing a Twitter account using its API. The metadata can differ between input and output bindings. For example, to use Twitter as an input binding, you need to specify the text to search for in tweets using the `query` field. Every time a matching tweet is sent, the Dapr sidecar will invoke the `/twitter-mention` endpoint on the service. It will also deliver the contents of the tweet.
```

A binding can be configured for input, output, or both. Interestingly, the binding doesn't explicitly specify input or output configuration. Instead, the direction is inferred by the usage of the binding along with configuration values.

> The [Dapr documentation for resource bindings](https://docs.dapr.io/operations/components/setup-bindings/supported-bindings/) provides a complete list of the available bindings and their specific configuration settings.

Reviewer: Move this to references because now we've already linked to this page before from the text

Reviewer 2: Cannot find the earlier reference. In the updated code, I see it at line 170. Then lines 189 and 193 reference specific bindings.

Reviewer: In the latest version, there are references on line 66 and 162. They both refer to [1] in the ##References section.
Technically, this is a good way to solve it because we only need to specify the URL once in the text. I'm not sure if Docs team wants us to reference the same page from multiple points in the text though. There's been a review comment on it in another chapter.

### Cron binding

Pay close attention to Dapr's Cron binding. It doesn't subscribe to events from an external system. Instead, this binding uses a configurable interval schedule to trigger your application. The binding provides a simple way to implement a background worker to wake up and do some work at a regular interval, without the need to implement an endless loop with a configurable delay. Here's an example of a Cron binding configuration:

```yaml
apiVersion: dapr.io/v1alpha1
kind: Component
metadata:
  name: checkOrderBacklog
  namespace: default
spec:
  type: bindings.cron
  metadata:
  - name: schedule
    value: "@every 30m"
```

In this example, Dapr triggers a service by invoking the `/checkOrderBacklog` endpoint every 30 minutes. There are several patterns available for specifying the `schedule` value. For more information, see the [Cron binding documentation](ocs.dapr.io/operations/components/setup-bindings/supported-bindings/cron/).

## Reference architecture: eShopOnDapr

> The accompanying eShopOnDapr reference application implements an output binding example. It triggers the Dapr [SendGrid](https://docs.dapr.io/operations/components/setup-bindings/supported-bindings/sendgrid/) binding to send a user an email when a new order is started. You can find this binding in the `eshop-email.yaml` file in the components folder:

Reviewer: ```suggestion
The accompanying eShopOnDapr reference application implements an output binding example. It triggers the Dapr [SendGrid](https://docs.dapr.io/operations/components/setup-bindings/supported-bindings/sendgrid/) binding to send a user an email when a new order is placed. You can find this binding in the `eshop-email.yaml` file in the components folder:
```

```yaml
apiVersion: dapr.io/v1alpha1
kind: Component
metadata:
  name: sendmail
  namespace: default
spec:
  type: bindings.twilio.sendgrid
  metadata:
  - name: apiKey
    secretKeyRef:
      name: sendGridAPIKey
auth:
  secretStore: eshop-secretstore
```

This configuration uses the [Twilio SendGrid](https://github.com/dapr/components-contrib/tree/master/bindings/twilio) binding component. Note how the API key for connecting to the service consumes a Dapr secret reference. This approach keeps secrets outside of the configuration file. Read the [Dapr secrets management chapter](secrets-buildingblocks.md) to learn more about Dapr secrets.

> The binding specifies a binding component that can be invoked using the `/sendmail` endpoint on the Dapr sidecar. Here's a code snippet in which an email is sent whenever an order is started:

Reviewer: ```suggestion
The binding configuration specifies a binding component that can be invoked using the `/sendmail` endpoint on the Dapr sidecar. Here's a code snippet in which an email is sent whenever an order is started:
```

```csharp
public async Task Handle(OrderStartedDomainEvent notification, CancellationToken cancellationToken)
{
    var payload = new
    {
        metadata = new
        {
            emailFrom = "eShopOn@dapr.io",
            emailTo = notification.UserName,
            subject = $"Your eShopOnDapr order #{notification.Order.Id}",
        },
        data = CreateEmailBody(notification),
        operation = "create"
    };

    var content = new StringContent(
        JsonSerializer.Serialize(payload),
        Encoding.UTF8, "application/json");

    var httpClient = _clientFactory.CreateClient();

>     var response = await httpClient.PostAsync($"http://localhost:{_daprHttpPort}/v1.0/bindings/sendmail", content);

Reviewer: Use the .NET SDK instead.

}
```

> > [!NOTE]

Reviewer: Can be removed when we use the SDK
